[Fix GCP plugin] Introduce public GCE Bootstrap Image

tewfik-ghariani opened the pull request:
### Background

After reading some docs related to running NixOs on GCE, I've seen that we currently have to bootstrap our own image from an object publicly shared in GS according to this : https://nixos.wiki/wiki/Install_NixOS_on_GCE

And at this time, the nixops-gcp plugin is deemed broken specifically due to this particular 'bootstrap-image' resource

### Proposed Solution

However, according to the GCP docs, we can actually just bake some images in a given GCP account, and then make the images public. 
https://cloud.google.com/compute/docs/images/managing-access-custom-images#share-images-publicly

Even better, we may create image families to allow better flexibility to manage single images. No need to specify the machine name per say but just the family. And under the hood, the maintainer may deprecate old ones and keep the 'latest' up-to-date.

### Validation

I started by testing the whole scenario using gcloud commands and it worked like a charm!

_Building an image from source_

```
$ gcloud compute images create nixos-18091228a4c4cbb613c-x86-64-linux  \
              --source-uri gs://nixos-cloud-images/nixos-image-18.09.1228.a4c4cbb613c-x86_64-linux.raw.tar.gz \
              --family=nixos-1809
```
_Making the image public_

```
$ gcloud compute images  add-iam-policy-binding nixos-18091228a4c4cbb613c-x86-64-linux \
             --member='allAuthenticatedUsers'                                                                              \
             --role='roles/compute.imageUser' 
```

And then that image may be used publicly so that all nixops users won't have the need to provision their own 'bootstrap-image' resource for every deployment. 

 ```
$ gcloud compute instances create test-nixos-18  \
             --image-family=nixos-1809                        \
             --zone=europe-west1-c                             \
             --image-project=predictix-operations
```

### Implementation
Implemented the change described above in this codebase and the gce plugins should be working fine now. 

#### The way it works 

- Extract the public image based on family name and project
- Create a volume out of that image [ No need to copy it in every deployment]
- Proceed naturally and use that volume as bootDisk


#### A summary of what has been achieved overall

- Deprecated the automatic provisioning of gce-image resource.
- Introduced bootstrap images as public image families.
- Added 'publicImageProject' option for blockDeviceMapping and gceDisk to specify the project from which publicly available image families are retrieved from.
- Controlled the exceptions and error messages


#### Small Additions
- Display GCE volume size in 'nixops-info'
- Remove Uniqueness condition on fileSystemsOption
- Update some GCP documentation links

#### *ToDo*

- Update the 'image-family' to point to the 'gce-images.nix' file under nixpkgs
```  
  image-family = import ./gce-images.nix;
  # To be changed to
  # image-family = import <nixpkgs/nixos/modules/virtualisation/gce-images.nix>;
```
- Prepare the images in the nixOs GCP account : https://github.com/NixOS/nixos-org-configurations/issues/110
- Update the 'project' variable in the 'gce-images.nix'

-------

Please let me know what do you think about this. If you have any suggestions or recommendations, feel free to share them.
cc @PsyanticY @AmineChikhaoui 


cc @rbvermaa Can we consider this as part of https://github.com/NixOS/nixpkgs/issues/6991

PsyanticY commented on nix/gce.nix:
```diff
@@ -193,7 +200,7 @@ let
     options = {
       gce = mkOption {
         default = null;
-        type = with types; uniq (nullOr (submodule gceDiskOptions));
```
why remove uniq ?

tewfik-ghariani replied:
So that we don't find ourselves in a situation where we have to re-define all options 
**Example**
*gce.nix*
```
      fileSystems."/data" = {
        fsType = "xfs";
        options = [ "noatime" "nodiratime" ];
        autoFormat = true;
        formatOptions = "-f";
        gce.disk = null;
        gce.disk_name = null;
        gce.size = lib.mkDefault 500;
        gce.diskType = "ssd";
        gce.encrypt = encrypt;
      };
```
*partitions-gce.nix*
```
 {
   fileSystems."/data".gce.disk      = lib.mkForce resources.gceDisks."${machine}-data";
   fileSystems."/data".gce.disk_name = lib.mkForce "data";
 } 
```
This error is raised 

```
error: The unique option `fileSystems./data.gce' is defined multiple times, in `<unknown-file>' and `/nixops/partitions-gce.nix'.
(use '--show-trace' to show detailed location information)
```
_Solution_

```
      fileSystems."/data".gce = lib.mkOverride 10
      {
        disk      = resources.gceDisks."${machine}-data";
        disk_name = "data";
        size      = volumeSize;
        diskType  = "ssd";
        encrypt   = encrypt;
      };
```

adisbladis commented on nix/gce-images.nix:
```diff
@@ -0,0 +1,7 @@
+  "18.09" = "nixos-1809";
+
+  latest = self."18.09";
+  project = "predictix-operations";
```
This should be updated to a "trusted" upstream like nixos/nix-community.

tewfik-ghariani replied:
Definitely, added the steps to prepare the images in the description :

_Building an image from source_
```
$ gcloud compute images create nixos-18091228a4c4cbb613c-x86-64-linux  \
              --source-uri gs://nixos-cloud-images/nixos-image-18.09.1228.a4c4cbb613c-x86_64-linux.raw.tar.gz \
              --family=nixos-1809
```
_Making the image public_
```
$ gcloud compute images  add-iam-policy-binding nixos-18091228a4c4cbb613c-x86-64-linux \
             --member='allAuthenticatedUsers'                                                                              \
             --role='roles/compute.imageUser'  
```
@AmineChikhaoui should be familiar with these operations

AmineChikhaoui replied:
I think we should update the `create-gce.sh` script in nixpkgs to add those steps. Thanks for researching this !
@adisbladis yeah we'll have to host this under a NixOS account. I'm looking into that. Also we're missing a couple of releases in GCP images, so will start first with 20.03 and see if it works.

adisbladis commented on nixopsgce/backends/gce.py:
```diff
@@ -66,7 +66,8 @@ def parse_block_device(xml):
                 'snapshot': self.get_option_value(xml, 'snapshot', str, optional = True),
-                'image': self.get_option_value(xml, 'image', 'resource', optional = True),
+                'image': self.get_option_value(xml, 'image', str, optional = True),
+                'publicImageProject' : self.get_option_value(xml, 'publicImageProject', str, optional = True),
```
What happens if this is `None`(null)?

tewfik-ghariani replied:
This must be optional because in case it has value of `None`, the old method would be used consisting in creating the disk from our own bootstrap image

```                # Retrieve GCENodeImage based on family name and project
                if v['publicImageProject']:
                    try:
                        img = self.connect().ex_get_image_from_family(
                                  image_family=v['image'],
                                  ex_project_list=[v['publicImageProject']],
                                  ex_standard_projects=False,
                              )
               .......................
                .........................
                else:
                    img = v['image']

grahamc commented on nix/gce.nix:
```diff
@@ -211,6 +218,12 @@ let
     };
   };
 
+  nixosVersion = builtins.substring 0 5 (config.system.nixos.version or config.system.nixosVersion);
```
Just noting this is very similar to how nixops-aws works: https://github.com/NixOS/nixops-aws/blob/master/nix/ec2.nix#L166

adisbladis commented:
@AmineChikhaoui Could you check this PR?

AmineChikhaoui commented on nix/gce-images.nix:
```diff
@@ -0,0 +1,7 @@
+let self = {
```
I think we can have this file in nixpkgs somhow like `ec2-amis.nix` and export an attrs:
```
{ imageFamily = "nixos-20-03";
  project = "nixos-org";
}
```
So that updates of images are automatic and without changes to nixops-gce.

AmineChikhaoui commented on nix/gce.nix:
```diff
@@ -211,6 +218,12 @@ let
+
+  image-family = import ./gce-images.nix;
+  # To be changed to
+  # image-family = import <nixpkgs/nixos/modules/virtualisation/gce-images.nix>;
```
oh you made a comment about it already :)
nitpick: image-family -> imageFamily

tewfik-ghariani replied:
Done https://github.com/nix-community/nixops-gce/pull/1/commits/fd88dea0b5a953bd15056778f95036dec0dfb2b5 , https://github.com/nix-community/nixops-gce/pull/4/commits/3322656bd49cc30103547729856f102142669768

AmineChikhaoui commented on nixopsgce/resources/gce_disk.py:
```diff
@@ -96,14 +97,44 @@ def create(self, defn, check, allow_reboot, allow_recreate):
                      .format(defn.size if defn.size else "auto", extra_msg))
+
+            # Retrieve GCENodeImage based on family name and project
+            if defn.public_image_project:
```
hm I wonder if there is benefit from creating disks out of images ? afaik root disks are created automatically in the machine resource

tewfik-ghariani replied:
True but I did not change the current logic. I think it won't harm to leave it just in case someone wanted to create extra gceDisks from an image ( public or not ) given that it's technically possible to do that in GCP

AmineChikhaoui commented on nix/gce.nix:
```diff
@@ -412,9 +425,15 @@ in
     nixpkgs.system = mkOverride 900 "x86_64-linux";
 
+    # Using NixOs public GCE images by default
+    deployment.gce.bootstrapImage = mkDefault (
```
I'm a bit confused by having 2 options `image` `bootstrapImage` do we still need both with this approach ?

tewfik-ghariani replied:
`image` is an option for the GCEdisk and `bootstrapImage` is an option for GCE deployment
GCE machines will always use one of the bootstrapImages provided in the '*gce-images.nix*' and their corresponding root disks will have that as ***image***

The use case of specifying an *image* from which an extra GCEDisk is created is explained in the previous comment. But it's up to you of course ;)

tewfik-ghariani commented:
Hello again :)) 

As a result of the meeting with @AmineChikhaoui  and based on his remarks, updated the code in a way to make it more intuitive to create root disks or separate disks from a public image.

The usage syntax shall be as simple as the following :

```nix
machine = {
  deployment.gce = {
    bootstrapImage = {    
      name = "base-image-bootstrap";
      family = null;
      project = "nixos-org";
    };
  };
}
```
```nix
resources.gceDisks.main-volume = {
  image = {
    name = null;
    family = "super-family";
    project = "another-project";
  };
};
```
 
I've completed my work based on the new changes related to types and options per #7 

_Change log_

- Added an example of Machines + GCE Disks.
- Introduced imageOptions and defined it as submodule.
> The imageOptions contains the following :
>  - image : either image-resource or the name of an existing image
>  - family : the image family to be used
>  - project : the parent project containing the image/imageFamily

> Note that we have to specify either name or family.
> Default value = {}
> Then adjusted the image option to inherit from the imageOptions submodule in
> - gceDiskOptions in gce.nix
> - gce-disk.nix
> Finally, updated the bootstrapImage gce option accordingly
- Created a new common retrieve_gce_image method to fetch the GCENodeImage object.
- Removed the publicImageProject attribute
- Upgraded libcloud
- Ignored libcloud mypy annotations

Full list of changes : https://github.com/nix-community/nixops-gce/compare/master...tewfik-ghariani:gcp-plugin

cc @AmineChikhaoui , @adisbladis  can you review please? Or do you prefer if I create a new PR?

tewfik-ghariani commented:
For some reason, the changes are no longer visible and the commit log has gotten confusing. Maybe it's because the second PR containing the original changes was already merged and this branch was re-based of master

In any case, I will be closing this PR and creating a new one

Cheers :))

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixops-flake-gce/22355/2
